acvm/src/pwg/directives.rs

```diff
@@ -15,13 +15,13 @@ use super::{get_value, insert_value, sorting::route, witness_to_value};
 pub fn solve_directives(
     initial_witness: &mut BTreeMap<Witness, FieldElement>,
     directive: &Directive,
-) -> Result<(), OpcodeResolutionError> {
+) -> Result<Option<Directive>, OpcodeResolutionError> {
```

Comment:
I don't think the solve should return an optional directive. The directive is either solved or not. If it is solved, then you can get the witness values from the witness map.

Reply:
Agreed - If a directive is not solved, then it means we did not find a witness that it needed, this will return an Error

Reply:
I was unsure on whether I should do this but I went with returning a directive mainly to remove duplication of code as the logging information needed by whomever is calling solve is very similar to the Log directive itself. Whomever is calling solve must check that we have a finalized string output and whether the log is a trace or a println. This can be seen in this draft PR: https://github.com/noir-lang/noir/pull/872. I will look into change this as I agree it confuses the idea of whether the directive has been solved or not.